src/Language/GraphQL/Draft/Parser.hs

```diff
@@ -53,6 +56,32 @@ runParser parser t =
 parseExecutableDoc :: Text -> Either Text AST.ExecutableDocument
 parseExecutableDoc = runParser executableDocument
 
+-- * Schema definition.
+
+-- | Parser for a 'AST.SchemaDefinition'.
+schemaDefinition :: Parser AST.SchemaDefinition
+schemaDefinition = tok "schema" *> braces (
+    AST.SchemaDefinition
+      <$> optional directives
+      <*> many1 rootOperationParser
+  )
+
+-- | Parser for a 'AST.RootOperationTypeDefinition'.
+rootOperationParser :: Parser AST.RootOperationTypeDefinition
+rootOperationParser =
+  AST.RootOperationTypeDefinition
+    <$> (operationTypeParser <* tok ":")
+    <*> (AST.NamedType <$> nameParser)
+
+-- | Parser for a 'AST.TypeSystemDefinition'.
+typeSystemDefinition :: Parser [AST.TypeSystemDefinition]
+typeSystemDefinition = whiteSpace *> (concat <$> many1 (
+      (\(AST.SchemaDocument d) -> AST.TypeSystemDefinitionType <$> d) <$> schemaDocument
+  <|> (: []) . AST.TypeSystemDefinitionSchema <$> schemaDefinition ))
```

Comment:
Do we need to consume `whiteSpace` explicitly as part of this? The majority of the parsing code doesn't need to do this (it consumes whitespace implicitly as needed), and the GraphQL spec doesn't say anything special about it for `TypeSystemDefinition`.

Again, it would be nice to follow the GraphQL spec literally here, parsing this as
```
TypeSystemDefinition :
   SchemaDefinition
   TypeDefinition
   DirectiveDefinition
```
For us this means that `typeSystemDefinition` would be defined as the disjunction of three other parsers. I don't think we have a parser for `DirectiveDefinition` yet.

The calculations being done here make me feel less confident the right thing is being defined. Is it necessary to use lambdas and the construction of empty lists here, rather than composing a `typeSystemDefinition` of parser combinators? If so, could you please add a comment what is being computed and why?

Reply:
Hi @abooij, thanks for the review! I applied some of the comments you mentioned here. I removed the `whiteSpace` as you requested but bear in mind that we had some issues in real world code where users would have an empty line before anything in the `graphql.schema` and it would break the parser. Also, the list constructor is needed here cause there can be **many** types inside a typeSystemDefinition. 😄

Reply:
> there can be many types inside a typeSystemDefinition

What do you mean by this?

Why does `typeSystemDefinition` parse a *list*, rather than simply `AST.TypeSystemDefinition`?

I do think it would be nice to have this parser code to look visually close to the spec. So I do think writing `typeSystemDefinition` using alternative/disjunctives `<|>` would be preferable.

(If `whiteSpace` is needed, it should be added, of course. But then we need to figure out why it's needed here, and nowhere else.)